### PR TITLE
raftstore: fix possible stale snap state

### PR DESCRIPTION
**Pull request by BusyJay**
ref #1373 

@siddontang @hhkbp2 PTAL

### CODE REVIEW

File: src/raftstore/store/peer_storage.rs [hunk 1]
```diff
@@ -396,43 +392,58 @@ impl PeerStorage {
...
+
+        if SnapState::Relax != *snap_state {
+            panic!("{} unexpected state: {:?}", self.tag, *snap_state);
         }
-        let task = RegionTask::Gen { region_id: self.get_region_id() };
+
+        if *tried_cnt >= MAX_SNAP_TRY_CNT {
+            let cnt = *tried_cnt;
+            *tried_cnt = 0;
+            return Err(raft::Error::Store(box_err!("failed to get snapshot after {} times", cnt)));
```
Comment (siddontang) [inline src/raftstore/store/peer_storage.rs]: seem that returning other error will panic in raft?

File: src/raftstore/store/peer_storage.rs [hunk 2]
```diff
@@ -396,43 +392,58 @@ impl PeerStorage {
...
+        }
+
+        info!("{} requesting snapshot...", self.tag);
+        *tried_cnt += 1;
+        let (tx, rx) = mpsc::sync_channel(1);
```
Comment (siddontang) [inline src/raftstore/store/peer_storage.rs]: do we need sync_channel here?

**Comment by siddontang** [comment]
Rest LGTM

File: src/raftstore/store/peer_storage.rs [hunk 3]
```diff
@@ -396,43 +392,58 @@ impl PeerStorage {
...
                   self.tag,
-                  snap_tried_cnt);
-            self.snap_tried_cnt.store(snap_tried_cnt, Ordering::Relaxed);
-            *snap_state = SnapState::Generating;
-        } else {
-            return Err(raft::Error::Store(raft::StorageError::SnapshotTemporarilyUnavailable));
+                  *tried_cnt);
+        }
+
+        if SnapState::Relax != *snap_state {
```
Comment (hhkbp2) [inline src/raftstore/store/peer_storage.rs]: What's the role of this line?

File: src/raftstore/store/peer_storage.rs [hunk 4]
```diff
@@ -396,43 +392,58 @@ impl PeerStorage {
...
+        let task = RegionTask::Gen {
+            region_id: self.get_region_id(),
+            notifier: tx,
+        };
         if let Err(e) = self.region_sched.schedule(task) {
             error!("{} failed to schedule task snap generation: {:?}",
                    self.tag,
                    e);
-            *snap_state = SnapState::Failed;
+            // update the status next time the function is called, also backoff for retry.
```
Comment (hhkbp2) [inline src/raftstore/store/peer_storage.rs]: Which code is this line of comment for?

**Comment by hhkbp2** [comment]
Rest LGTM

File: src/raftstore/store/peer_storage.rs (hunk 1 above)
Comment (BusyJay) [inline src/raftstore/store/peer_storage.rs]: Yes.

File: src/raftstore/store/peer_storage.rs (hunk 2 above)
Comment (BusyJay) [inline src/raftstore/store/peer_storage.rs]: So I can limit the buffer's size to 1.

File: src/raftstore/store/peer_storage.rs (hunk 3 above)
Comment (BusyJay) [inline src/raftstore/store/peer_storage.rs]: Check if the peer is in a good state. Other states like `Applying` is invalid and there should be bugs somewhere.

File: src/raftstore/store/peer_storage.rs (hunk 4 above)
Comment (BusyJay) [inline src/raftstore/store/peer_storage.rs]: None, it just explains why not update the state and retry here.

**Comment by siddontang** [comment]
LGTM
PTAL @hhkbp2

**Comment by BusyJay** [comment]
ping

**Comment by hhkbp2** [comment]
LGTM

**Pull request merged**